java/src/main/java/com/logicalclocks/base/constructor/QueryBase.java

```diff
@@ -0,0 +1,285 @@
+/*
+ *  Copyright (c) 2020-2022. Hopsworks AB
+ *
+ *  Licensed under the Apache License, Version 2.0 (the "License");
+ *  you may not use this file except in compliance with the License.
+ *  You may obtain a copy of the License at
+ *
+ *  http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *  Unless required by applicable law or agreed to in writing, software
+ *  distributed under the License is distributed on an "AS IS" BASIS,
+ *  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *
+ *  See the License for the specific language governing permissions and limitations under the License.
+ *
+ */
+
+package com.logicalclocks.base.constructor;
+
+import com.fasterxml.jackson.annotation.JsonIgnore;
+import com.logicalclocks.base.Feature;
+import com.logicalclocks.base.FeatureStoreException;
+import com.logicalclocks.base.Storage;
+import com.logicalclocks.base.engine.FeatureGroupUtils;
+import com.logicalclocks.base.metadata.FeatureGroupBase;
+
+import lombok.Getter;
+import lombok.NoArgsConstructor;
+import lombok.Setter;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.io.IOException;
+import java.text.ParseException;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Map;
+import java.util.stream.Collectors;
+
+@NoArgsConstructor
+public class QueryBase {
+
+  private static final Logger LOGGER = LoggerFactory.getLogger(QueryBase.class);
+
+  @Getter
+  @Setter
+  private FeatureGroupBase leftFeatureGroup;
+  @Getter
+  @Setter
+  private List<Feature> leftFeatures;
+  @Getter
+  @Setter
+  private Long leftFeatureGroupStartTime;
+  @Getter
+  @Setter
+  private Long leftFeatureGroupEndTime;
+  @Getter
+  @Setter
+  private List<Join> joins = new ArrayList<>();
+  @Getter
+  @Setter
+  private FilterLogic filter;
+  @Getter
+  @Setter
+  private Boolean hiveEngine = false;
+
+  private FeatureGroupUtils utils = new FeatureGroupUtils();
+
+  public QueryBase(FeatureGroupBase leftFeatureGroup, List<Feature> leftFeatures) {
+    this.leftFeatureGroup = leftFeatureGroup;
+    this.leftFeatures = leftFeatures;
+  }
+
+  public String sql() {
```

Comment:
this and other methods like this should be made abstract so that they are not left unassigned (multiple methods in this class are overwritten in Query.java)

Reply:
QueryBase is used in API and can't make abstract. In general I agree with you. But let's make this cleanup in version 2, as we need to make good abstract classes.

Reply:
OK its abstract class now